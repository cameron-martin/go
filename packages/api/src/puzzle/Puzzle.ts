export type BoardPosition = readonly [number, number];

export type PlayResult =
  | { type: 'wrong'; response?: BoardPosition }
  | { type: 'correct' }
  | { type: 'continue'; response: BoardPosition };

const positionEquals = (position1: BoardPosition, position2: BoardPosition) =>
  position1[0] === position2[0] && position1[1] === position2[1];

export class Puzzle {
  static create(spec: PuzzleSpec) {
    return new Puzzle(spec);
  }

  readonly initialStones: InitialStones;
  readonly sequences: PuzzleTree[];

  private constructor(spec: PuzzleSpec) {
    this.initialStones = spec.initialStones;
    this.sequences = spec.sequences;
  }

  get spec(): PuzzleSpec {
    return { initialStones: this.initialStones, sequences: this.sequences };
  }

  playSequence(positions: BoardPosition[]): PlayResult {
    let sequences = this.sequences;

    for (let i = 0; i < positions.length; i++) {
      const position = positions[i];

      const sequence = sequences.find((sequence) =>
        positionEquals(sequence.position, position),
      );

      if (!sequence) {
        return {
          type: 'wrong',
        };
      }

      if (sequence.type === 'leaf') {
        if (sequence.outcome === 'correct') {
          return { type: 'correct' };
        }

        if (sequence.outcome === 'wrong') {
          return { type: 'wrong', response: sequence.response };
        }
      } else if (i === positions.length - 1) {
        return { type: 'continue', response: sequence.response };
      } else {
        sequences = sequence.children;
      }
    }

    throw new Error('Sequence cannot be empty');
  }

  get area(): { min: BoardPosition; max: BoardPosition } {
    const min: [number, number] = [Infinity, Infinity];
    const max: [number, number] = [-Infinity, -Infinity];

    const processPosition = (stone: BoardPosition) => {
      min[0] = Math.min(min[0], stone[0]);
      min[1] = Math.min(min[1], stone[1]);
      max[0] = Math.max(max[0], stone[0]);
      max[1] = Math.max(max[1], stone[1]);
    };
    this.initialStones.you.forEach(processPosition);
    this.initialStones.computer.forEach(processPosition);
    this.sequences.forEach(function processNode(node) {
      processPosition(node.position);
      if (node.response) {
        processPosition(node.response);
      }

      if (node.type === 'branch') {
        node.children.forEach(processNode);
      }
    });

    return { min, max };
  }
}

export interface InitialStones {
  you: BoardPosition[];
  computer: BoardPosition[];
}

export interface PuzzleSpec {
  sequences: PuzzleTree[];
  initialStones: InitialStones;
}

export type PuzzleTree = PuzzleBranch | PuzzleLeaf;

export interface PuzzleBranch {
  type: 'branch';
  position: BoardPosition;
  response: BoardPosition;
  children: PuzzleTree[];
}

export interface PuzzleLeaf {
  type: 'leaf';
  position: BoardPosition;
  response?: BoardPosition;
  outcome: Outcome;
}

export type Outcome = 'correct' | 'wrong';
