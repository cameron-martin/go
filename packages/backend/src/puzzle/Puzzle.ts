export enum Player {
  You,
  Computer,
}

export type BoardPosition = readonly [number, number];

export type PlayResult =
  | { type: 'wrong'; response?: BoardPosition }
  | { type: 'correct' }
  | { type: 'continue'; response: BoardPosition };

const positionEquals = (position1: BoardPosition, position2: BoardPosition) =>
  position1[0] === position2[0] && position1[1] === position2[1];

export class Puzzle {
  static create(spec: PuzzleSpec) {
    return new Puzzle(spec);
  }

  readonly initialStones: InitialStones;
  readonly sequences: PuzzleTree[];

  private constructor(spec: PuzzleSpec) {
    this.initialStones = spec.initialStones;
    this.sequences = spec.sequences;
  }

  playSequence(positions: BoardPosition[]): PlayResult {
    let sequences = this.sequences;

    for (let i = 0; i < positions.length; i++) {
      const position = positions[i];

      const sequence = sequences.find(sequence =>
        positionEquals(sequence.position, position),
      );

      if (!sequence) {
        return {
          type: 'wrong',
        };
      }

      if (sequence.type === 'leaf') {
        if (sequence.outcome === 'correct') {
          return { type: 'correct' };
        }

        if (sequence.outcome === 'wrong') {
          return { type: 'wrong', response: sequence.response };
        }
      } else if (i === positions.length - 1) {
        return { type: 'continue', response: sequence.response };
      } else {
        sequences = sequence.children;
      }
    }

    throw new Error('Sequence cannot be empty');
  }
}

export interface InitialStones {
  [Player.You]: BoardPosition[];
  [Player.Computer]: BoardPosition[];
}

export interface PuzzleSpec {
  sequences: PuzzleTree[];
  initialStones: InitialStones;
}

export type PuzzleTree = PuzzleBranch | PuzzleLeaf;

export interface PuzzleBranch {
  type: 'branch';
  position: BoardPosition;
  response: BoardPosition;
  children: PuzzleTree[];
}

export interface PuzzleLeaf {
  type: 'leaf';
  position: BoardPosition;
  response?: BoardPosition;
  outcome: Outcome;
}

export type Outcome = 'correct' | 'wrong';
