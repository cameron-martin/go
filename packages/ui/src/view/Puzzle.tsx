import React, { useState, useCallback, useEffect } from 'react';
import { Button, makeStyles, Typography, Container } from '@material-ui/core';
import { either } from 'fp-ts';
import { ApiClient } from '@tsumego/api-client';
import Board, { BoardCrop } from './board/Board';
import { GoGame, BoardPosition } from '../model/GoGame';
import PuzzleInstructions from './PuzzleInstructions';
import Loading from './Loading';

interface Props {
  apiClient: ApiClient;
}

type Game =
  | Readonly<{ loadState: 'pending' }>
  | Readonly<{ loadState: 'rejected' }>
  | Readonly<{
      loadState: 'fulfilled';
      moveState: 'computers-turn' | 'humans-turn' | 'correct' | 'wrong';
      sequence: readonly BoardPosition[];
      area: BoardCrop;
      id: string;
      game: GoGame;
    }>;

const computerPlayer = 'white';
const humanPlayer = 'black';

const useStyles = makeStyles({
  root: {
    flex: '1 0 auto',
    display: 'flex',
    alignItems: 'center',
    justifyContent: 'center',
    flexDirection: 'column',
  },
});

export default function Puzzle({ apiClient }: Props) {
  const classes = useStyles();

  const [gameState, setGameState] = useState<Game>({
    loadState: 'pending',
  });

  const loadPuzzle = useCallback(async () => {
    setGameState({
      loadState: 'pending',
    });

    try {
      const { id, initialStones, area } = await apiClient.puzzle.getRandom();

      setGameState({
        loadState: 'fulfilled',
        moveState: 'humans-turn',
        id,
        sequence: [],
        game: GoGame.create(19, {
          [computerPlayer]: initialStones.computer,
          [humanPlayer]: initialStones.you,
        }),
        area: {
          min: [area.min[0] - 2, area.min[1] - 2],
          max: [area.max[0] + 2, area.max[1] + 2],
        },
      });
    } catch (err) {
      setGameState({
        loadState: 'rejected',
      });

      throw err;
    }
  }, [apiClient.puzzle]);

  useEffect(() => {
    loadPuzzle();
  }, [loadPuzzle]);

  useEffect(() => {
    if (
      gameState.loadState === 'fulfilled' &&
      gameState.moveState === 'computers-turn'
    ) {
      apiClient.puzzle
        .solve(gameState.id, gameState.sequence)
        .then((response) => {
          if (response.type === 'continue') {
            setGameState({
              ...gameState,
              moveState: 'humans-turn',
              game: gameState.game.playValidMove({
                player: computerPlayer,
                position: response.response,
              }),
            });
          } else if (response.type === 'correct' || response.type === 'wrong') {
            setGameState({ ...gameState, moveState: response.type });
          }
        });
    }
  }, [gameState, apiClient.puzzle]);

  const playMove = useCallback((position: BoardPosition) => {
    setGameState((gameState) => {
      if (
        gameState.loadState === 'fulfilled' &&
        gameState.moveState === 'humans-turn'
      ) {
        const newGame = gameState.game.playMove({
          player: humanPlayer,
          position,
        });

        if (either.isLeft(newGame)) {
          return gameState;
        }

        return {
          ...gameState,
          sequence: gameState.sequence.concat([position]),
          game: newGame.right,
          moveState: 'computers-turn',
        };
      } else {
        return gameState;
      }
    });
  }, []);

  const handleNextPuzzle = useCallback(() => {
    setGameState({ loadState: 'pending' });

    loadPuzzle();
  }, [loadPuzzle]);

  if (gameState.loadState === 'pending') {
    return <Loading />;
  }

  if (gameState.loadState === 'rejected') {
    return (
      <div className={classes.root}>
        <Typography variant="body1" gutterBottom>
          Failed to load puzzle
        </Typography>
        <Button variant="contained" color="primary" onClick={loadPuzzle}>
          Retry
        </Button>
      </div>
    );
  }

  return (
    <Container maxWidth="sm">
      <PuzzleInstructions
        humanPlayer={humanPlayer}
        state={
          gameState.moveState === 'correct' || gameState.moveState === 'wrong'
            ? gameState.moveState
            : 'in-progress'
        }
        onNextPuzzle={handleNextPuzzle}
      />
      <Board game={gameState.game} playMove={playMove} crop={gameState.area} />
    </Container>
  );
}
