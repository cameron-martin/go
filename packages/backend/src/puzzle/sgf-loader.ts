import { PuzzleSpec, InitialStones, Player } from './Puzzle';
import {
  GameTree,
  Node,
  tryParseSgf,
  PropertyType,
  Property,
  Point,
} from './sgf-parser';

export function loadSgf(sgf: string): PuzzleSpec {
  const trees = tryParseSgf(sgf);

  if (trees.length !== 1) throw new Error('SGF must have one gametree');

  const tree = trees[0];

  if (tree.nodes.length === 0) throw new Error('SGF has no root nodes');

  const { initialStones, remainingTree } = getInitialStones(tree);

  return {
    initialStones,
    sequences: [],
  };
}

function loadGameTree(tree: GameTree) {}

function findKnownProperty<K extends PropertyType>(node: Node, type: K) {
  return node.properties.find(
    prop => prop.type === 'known' && prop.ident === type,
  ) as (Extract<Property, { type: 'known'; ident: K }> | undefined);
}

function expandPointComposition([upperLeft, lowerRight]: [Point, Point]) {
  const expanded: Point[] = [];

  for (let i = upperLeft[0]; i <= lowerRight[0]; i++) {
    for (let j = upperLeft[1]; j <= lowerRight[1]; j++) {
      expanded.push([i, j]);
    }
  }

  return expanded;
}

function expandCompressedPointList(list: Array<[Point, Point] | Point>) {
  return list.flatMap(point => {
    if (typeof point[0] === 'number') {
      return [point as Point];
    }

    return expandPointComposition(point as [Point, Point]);
  });
}

function getInitialStones(
  tree: GameTree,
): { initialStones: InitialStones; remainingTree: GameTree } {
  const { nodes, gameTrees } = tree;

  const [firstNode, ...remainingNodes] = nodes;

  const ab = findKnownProperty(firstNode, 'AB');
  const aw = findKnownProperty(firstNode, 'AW');

  const whitePoints = aw ? expandCompressedPointList(aw.value) : [];
  const blackPoints = ab ? expandCompressedPointList(ab.value) : [];

  // TODO: Work out which way round the players are

  return {
    initialStones: {
      [Player.You]: blackPoints,
      [Player.Computer]: whitePoints,
    },
    remainingTree: {
      nodes: remainingNodes,
      gameTrees,
    },
  };
}
