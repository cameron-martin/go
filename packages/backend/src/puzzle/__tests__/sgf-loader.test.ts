/**
 * @jest-environment node
 */

import { promises as fs } from 'fs';
import path from 'path';
import { loadSgf } from '../sgf-loader';
import { PuzzleSpec, Player } from '../Puzzle';

test('snapshots', async () => {
  const file = await fs.readFile(
    path.join(
      __dirname,
      'sgfs',
      'puzzles',
      'CD 1 - FAMOUS TSUMEGO COMPOSERS',
      'CHO CHIKUN Encyclopedia Life And Death - Elementary',
      'prob0001.sgf',
    ),
    'utf8',
  );

  const expected: PuzzleSpec = {
    initialStones: {
      [Player.Computer]: [[3, 0], [0, 1], [1, 1], [2, 1], [3, 1]],
      [Player.You]: [[4, 1], [5, 1], [1, 2], [2, 2], [3, 2], [1, 4]],
    },
    sequences: [
      { position: [1, 0], outcome: 'correct' },
      { position: [2, 0], outcome: 'wrong' },
      { position: [0, 0], outcome: 'wrong' },
      { position: [0, 1], outcome: 'wrong' },
      { position: [4, 0], outcome: 'wrong' },
    ],
  };

  expect(loadSgf(file)).toEqual(expected);
});
