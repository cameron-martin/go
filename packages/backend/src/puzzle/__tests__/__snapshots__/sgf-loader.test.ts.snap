// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`snapshots 1`] = `undefined`;
